accepted/00070-akka.md

```diff
@@ -0,0 +1,141 @@
+ - con: unexpected Taskomatic errors may bring Tomcat down as well. At the moment this does not happen, but as we rely on Taskomatic much more than in original Spacewalk, in many cases we already end up with a non-functioning Server anyway
+ - pro: currently Taskomatic runs as `root`, but not all tasks really require `root` privileges. Attack surface can thus be reduced by running code with Tomcat privileges by default
+ - con: Taskomatic tasks that do require root privileges will need some different permissions to run. One way is to use Akka Clustering (described below) with two processes (one `root` and one non-`root`) on the same machine
+
+### Leverage Akka Clustering to offload computation onto other machines
+
+Akka has an optional module called Clustering that allows to transparently move computation across processes (on the same machine or on different machines in a network). It is thus possible to instantiate Actors "remotely", and Akka takes care of serializing, delivering and deserializing Messages to them transparently.
+
+This has been demoed as proof-of-concept [in the HackWeek PR](https://github.com/uyuni-project/uyuni/pull/1903) (commits after February 15), and it is an interesting door to have opened by the work in this PR.
+
+Before considering this for the main codebase, the following aspects should be evaluated:
+
+ - out of all the workloads migrated to Akka, which ones make sense to distribute? In many cases, Uyuni bottlenecks lie in the database or Salt components, and those would not be affected by such a change
+ - does the trade-off make sense at all from a performance perspective? Addding a network layer introduces latency which might end up negating any speedup gained because of the cluster
+ - how can Akka Cluster nodes connect back to the one only Salt and database instances? During HackWeek, a simple SSH tunnel was used for demonstration's sake. Distributing Salt or the database are separate efforts with their own set of trade-offs
+ - is the extra administration overhead worth it? In which cases? Maintaining a cluster is much more work-intensive than maintaining a single VM
+ - how can we deal with upgrades? Akka Clustering has primitives for versioning, but making them work with existing packaging and delivery processes requires some study
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+ - as with all refactorings, we risk introducing regressions at least initially
+ - Akka itself is a complicated artifact. Although external quality seemed very good during HackWeek (no bugs or strange/undocumented behavior was noticed, great documentation) there is always a chance some internal bug hits us, and might be difficult to solve
+ - we get to maintain a packaged version of Akka
+ - performance has to be measured - we expect it to improve but that is not guaranteed
```

Comment:
Can we already define some metrics to measure the performance before and after the Akka introduction?

Reply:
One possible idea is to measure onboarding time and upgrade time of a given number of minions without and with the patch, as those operations exercise the Salt event code path and MessageQueue heavily.

I'd go incrementally though, creating specific benchmark for each migrated subsystem (before moving on to the next!)

Do you feel like adding some text in that direction is necessary, or maybe do you have any other suggestion?

Reply:
In the Motivation section, you point out that you are aiming at - among the other aspects - to performance.
So, IMHO it makes sense to define a rough idea of what would be the figure of merit for every aspect you want to improve (or at least the one with the heaviest impact, like performance). No need to be super detailed, onboarding time and upgrade time are already OK in my opinion.

Reply:
Right, added, thanks